storm-server/src/main/java/org/apache/storm/daemon/supervisor/SupervisorUtils.java

```diff
@@ -92,8 +92,13 @@ public static String getNumaIdForPort(Integer port, Map<String, Object> supervis
     }
 
     public static List<Integer> getSlotsPorts(Map<String, Object> supervisorConf) {
-        List<Integer> slotsPorts = (List<Integer>) supervisorConf.getOrDefault(DaemonConfig.SUPERVISOR_SLOTS_PORTS,
+        List<Integer> slotsPorts = new ArrayList<>();
+        List<Number> ports = (List<Number>) supervisorConf.getOrDefault(DaemonConfig.SUPERVISOR_SLOTS_PORTS,
                 new ArrayList<>());
+        for (Number port : ports) {
+            slotsPorts.add(port.intValue());
```

Comment:
This would may involve rounding or truncation which isn't really what we wanted.  

A port by definition is no larger than 65535. So logically it will be Integers.

There is an Integer check here https://github.com/apache/storm/blob/6e3707242e0e6835085d358bec0abee4f62b1426/storm-server/src/main/java/org/apache/storm/DaemonConfig.java#L742-L743. Not sure why it was not properly enforced.

Reply:
This is similar to how the code used to be (see https://github.com/apache/storm/blame/69e69d40d5c700465b49f4c23312f974fb2ca1f6/storm-server/src/main/java/org/apache/storm/daemon/supervisor/ReadClusterState.java#L101-L103) before this change: https://github.com/apache/storm/pull/3267/files

Reply:
How do we reproduce this? 

I changed the port to be a value larger than Integer.MAX_VALUE, the supervisor fails to start

```
2020-12-21 22:50:47.009 o.a.s.v.ConfigValidation main [INFO] Will use [class org.apache.storm.DaemonConfig, class org.apache.storm.Config] for validation
2020-12-21 22:50:47.250 o.a.s.u.Utils main [ERROR] Received error in thread main.. terminating server...
java.lang.Error: java.lang.IllegalArgumentException: Field SUPERVISOR_SLOTS_PORTS list entry must be an Integer within type range.
        at org.apache.storm.utils.Utils.handleUncaughtException(Utils.java:664) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.utils.Utils.handleUncaughtException(Utils.java:668) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.utils.Utils.lambda$createDefaultUncaughtExceptionHandler$2(Utils.java:1048) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1057) [?:1.8.0_262]
        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1052) [?:1.8.0_262]
        at java.lang.Thread.dispatchUncaughtException(Thread.java:1959) [?:1.8.0_262]
Caused by: java.lang.IllegalArgumentException: Field SUPERVISOR_SLOTS_PORTS list entry must be an Integer within type range.
        at org.apache.storm.validation.ConfigValidation$IntegerValidator.validateInteger(ConfigValidation.java:415) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.validation.ConfigValidation$IntegerValidator.validateField(ConfigValidation.java:401) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.validation.ConfigValidation$ListEntryCustomValidator.validateField(ConfigValidation.java:588) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.validation.ConfigValidation$ListEntryCustomValidator.validateField(ConfigValidation.java:602) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.validation.ConfigValidation.validateField(ConfigValidation.java:167) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.validation.ConfigValidation.validateFields(ConfigValidation.java:215) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.validation.ConfigValidation.validateFields(ConfigValidation.java:189) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.utils.ConfigUtils.readStormConfigImpl(ConfigUtils.java:440) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.utils.ConfigUtils.readStormConfig(ConfigUtils.java:166) ~[storm-client-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.daemon.supervisor.Supervisor.<init>(Supervisor.java:126) ~[storm-server-2.3.0.y.jar:2.3.0-SNAPSHOT]
        at org.apache.storm.daemon.supervisor.Supervisor.main(Supervisor.java:200) ~[storm-server-2.3.0.y.jar:2.3.0-SNAPSHOT]
```

Reply:
Without this change:

./bin/storm supervisor -c supervisor.slots.ports="[6700]"
2020-12-28 14:06:38.860 o.a.s.d.s.Supervisor main [ERROR] Failed to start supervisor

java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
    at org.apache.storm.daemon.supervisor.ReadClusterState.<init>(ReadClusterState.java:101) ~[storm-server-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
    at org.apache.storm.daemon.supervisor.Supervisor.launch(Supervisor.java:310) ~[storm-server-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
    at org.apache.storm.daemon.supervisor.Supervisor.launchDaemon(Supervisor.java:340) [storm-server-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
    at org.apache.storm.daemon.supervisor.Supervisor.main(Supervisor.java:201) [storm-server-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]

Reply:
Thanks